### PR TITLE
Added testing infrastructure

### PR DESCRIPTION
**Pull request by theoden8**
Added testing infrastructure based on random samples. The testing consists of two parts: unit tests, which use each function on the interface separately, and expression tests, which tests multiple functions at once (for precision, atomicity, mutability and other issues).

### CODE REVIEW

File: include/interface/blas1_interface_sycl.hpp [hunk 1]
```diff
@@ -436,12 +436,9 @@ size_t _iamin(Executor<ExecutorType> ex, int _N, vector_view<T, ContainerT> _vx,
  */
 template <typename T>
 void _rotg(T &_alpha, T &_beta, T &_cos, T &_sin) {
-  T abs_alpha = std::abs(_alpha);
-  T abs_beta = std::abs(_beta);
-  T roe = (abs_alpha > abs_beta) ? _alpha : _beta;
-  T scale = abs_alpha + abs_beta;
-  T norm;
-  T aux;
+  T abs_alpha = std::abs(_alpha), abs_beta = std::abs(_beta),
```
Comment (mehdi-goli) [inline include/interface/blas1_interface_sycl.hpp]: Do not change the code style to this style.

File: include/interface/blas1_interface_sycl.hpp [hunk 2]
```diff
@@ -462,8 +459,7 @@ void _rotg(T &_alpha, T &_beta, T &_cos, T &_sin) {
       aux = constant<T, const_val::one>::value;
     }
   }
-  _alpha = norm;
-  _beta = aux;
+  _alpha = norm, _beta = aux;
```
Comment (mehdi-goli) [inline include/interface/blas1_interface_sycl.hpp]: Do not change the code style to this style

File: tests/simple_axpy_test.cpp [hunk 3]
```diff
@@ -26,17 +26,10 @@ int main(int argc, char *argv[]) {
   }
   if (returnVal == 0) {
     // CREATING DATA
-    std::vector<double> vX(sizeV);
-    std::vector<double> vY(sizeV);
+    std::vector<double> vX(sizeV, 1.);
```
Comment (mehdi-goli) [inline tests/simple_axpy_test.cpp]: is there any reason that you set the value to one?

File: test/blas1_test.hpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+// the test class itself
+template <class B> class BLAS1_Test;
+
+template <class T_, template <class... As> class C_, class E_>
+class BLAS1_Test<blas1_test_args<T_, C_, E_>> : public ::testing::Test {
+ public:
+  using T = T_;
+  template <class U = T>
+  using C = C_<U>;
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: Convert all the name to the meaningful name

File: test/blas1_test.hpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+};
+// A "using" shortcut for the struct
+template <class T, template <class... As> class C = std::vector, class E = SYCL>
+using blas1_test_args = blas_templ_struct<T, C, E>;
+
+// the test class itself
+template <class B> class BLAS1_Test;
+
+template <class T_, template <class... As> class C_, class E_>
+class BLAS1_Test<blas1_test_args<T_, C_, E_>> : public ::testing::Test {
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: convert all the template name to the meaningful name

File: test/blas1_test.hpp [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+// We will treat template-specialized blas_templ_struct as a single class
+template <class T, template <class... As> class C, class E>
+struct blas_templ_struct {
+  using type = T;
+  template <class U = T>
+  using container = C<U>;
+  using executor = E;
+};
+// A "using" shortcut for the struct
+template <class T, template <class... As> class C = std::vector, class E = SYCL>
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: rename all the type name to the meaningful name

File: test/blas1_test.hpp [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+template <typename C>
+struct option_size;
+#define RANDOM_SIZE UINT_MAX
+#define REGISTER_SIZE(size, test_class)         \
+  template <>                                   \
+  struct option_size<class test_class> {        \
+    static constexpr const size_t value = size; \
+  };
+template <class T, typename C>
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: rename all the template parameter name to the meaningful name

File: test/blas1_test.hpp [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <interface/blas1_interface_sycl.hpp>
+
+using namespace cl::sycl;
+using namespace blas;
+
+template <typename C>
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: rename the template parameter name to the meaningful name

File: test/blas1_test.hpp [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+struct option_size;
+#define RANDOM_SIZE UINT_MAX
+#define REGISTER_SIZE(size, test_class)         \
+  template <>                                   \
+  struct option_size<class test_class> {        \
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: inconsistency with the type name. In line 31 test_class is capital case and here is lower case

File: test/blas1_test.hpp [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    static constexpr const size_t value = size; \
+  };
+template <class T, typename C>
+struct option_prec;
+#define REGISTER_PREC(type, val, Test_Class)   \
+  template <>                                  \
+  struct option_prec<type, class Test_Class> { \
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: inconsistency with the type name. In line 24 test_class is lower case and here is capital case

File: test/blas1_test.hpp [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+/*
+ * Template arguments:
+ * T = (scalar) type
+ * C = container
+ * E = executor
+ */
+
+// Wraps the above arguments into one template parameter.
+// We will treat template-specialized blas_templ_struct as a single class
+template <class T, template <class... As> class C, class E>
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: convert all template parameter to a meaningful name

File: test/blas1_test.hpp [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+struct blas_templ_struct {
+  using type = T;
+  template <class U = T>
+  using container = C<U>;
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: convert all template parameter name to a meaningful name

File: test/blas1_test.hpp [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+  template <class U = T>
+  static size_t rand_size() {
+    size_t ret = rand() >> 5;
+    int type_size = sizeof(U) * CHAR_BIT - std::numeric_limits<U>::digits10 - 2;
+    return (ret & (std::numeric_limits<size_t>::max() +
+                   (size_t(1) << (type_size - 2)))) +
+           1;
+  }
+
+  template <class U = T>
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: convert all template parameter names to  meaningful names

File: test/blas1_test.hpp [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+  static void set_rand(C<U> &vec, std::pair<U, U> &bounds) {
+    for (auto &x : vec) {
+      x = U(rand() % int(bounds.first - bounds.second) * 1000) * .001 -
+          bounds.second;
+    }
+  }
+
+  template <class U = T>
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: convert all template parameter names to  meaningful names

File: test/blas1_test.hpp [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+  static C<U> make_randcont(size_t size, std::pair<U, U> &&bounds = {-1, 1}) {
+    C<U> container(size);
+    set_rand(container, bounds);
+    return container;
+  }
+
+  template <class U = T>
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: convert all template parameter names to  meaningful names

File: test/blas1_test.hpp [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+  static void print_cont(const C<U> &vec, std::string name = "vector") {
+    std::cout << name << ": ";
+    for (auto &e : vec) std::cout << e << " ";
+    std::cout << std::endl;
+  }
+
+  template <class U = T>
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: convert all template parameter names to  meaningful names

File: test/blas1_test.hpp [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+  static buffer<U, 1> make_buffer(C<U> &vec) {
+    return buffer<U, 1>(vec.data(), vec.size());
+  }
+
+  template <class U = T>
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: convert all template parameter names to  meaningful names

File: test/blas1_test.hpp [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+        std::cout << " An exception " << std::endl;
+      }
+    });
+  }
+};
+
+// it is important that all tests are run with the same test size
+// so each time we access this function within the same program, we get the same
+// randomly generated size
+template <class TestClass>
```
Comment (mehdi-goli) [inline test/blas1_test.hpp:129]: Inconsistent class name style

File: test/blas1_test.hpp [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+    first = false;
+    N = TestClass::rand_size();
+  }
+  return N;
+}
+
+// templates the container type of the test class
+// TestClass - e.g. BLAS1_Test
+// T - element (or scalar) type, by default is TestClass::T
+template <class TestClass, class T = typename TestClass::T>
```
Comment (mehdi-goli) [inline test/blas1_test.hpp]: Inconsistent typename style. Convert all template parameter names to  meaningful names

File: test/unittest/blas1_asum_test.cpp [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+REGISTER_PREC(float, 1e-4, asum_test)
+REGISTER_PREC(double, 1e-6, asum_test)
+REGISTER_PREC(long double, 1e-7, asum_test)
+
+B1_TEST(asum_test) {
+  UNPACK_PARAM(asum_test);
+  size_t size = TEST_SIZE;
+  T prec = TEST_PREC;
+
+  auto vX = TestClass::make_randcont(size), vR = Container<TestClass>(1, T(0));
```
Comment (mehdi-goli) [inline test/unittest/blas1_asum_test.cpp]: Inconsistent class naming style

File: include/interface/blas1_interface_sycl.hpp (hunk 1 above)
Comment (theoden8) [reply include/interface/blas1_interface_sycl.hpp]: Forgot to revert after debugging. Will change back.

File: tests/simple_axpy_test.cpp (hunk 3 above)
Comment (theoden8) [reply tests/simple_axpy_test.cpp]: It's a replacement for the following std::foreach.

**Pull request closed without merging**